Find non matching edges

DanielYankura opened the pull request:
## Reason
It is possible for non-matching edges to be created when using MOOSE's convert hexahedral to tetrahedral tool. This adds a new mesh diagnostic tool that can check for non-matching/intersecting edges.

## Design
The function compares each element to nearby elements. If they are close to each other, then the edges from each are checked for overlap. To check for overlap the shortest line that connects the two edges is calculated. If that line's length is less than the user specified tolerance then it's counted as overlapping and info about the location of the intersection is printed, along with the total number of intersecting edge pairs.

## Impact
Users can now include examine_non_matching_edges = INFO/WARNING/NO_CHECK as a mesh diagnostics option. Users can also specify overlap tolerance by including intersection_tol but this is not necessary.

closes #26596

Commit:
Basic outline for checkNonMatchingEdges

-Adds checkNonMatchingEdges function to MeshDiagnosticsGenerator.C/.h
-Includes basic outline of algorithm
-Includes nested for-loops for looping over all mesh elements

refs #26596

Adding in sudo-code for the algorithm that finds edge overlap

-Still need a method to get (x,y,z) coordinates from a node
-Also added the function to the header file

Updated method for finding intersecting edges

-Implemented method to find closest line connecting two edges
-If the line length is < tol it counts as intersecting
-Only checks is edges don't share any nodes
-Needs improvement to ensure it doesn't double check edges
-Needs improvement to ensure it loops over all edges even if they're in different blocks/volumes
-Need to improve tolerance so it's not hard-coded
-Need to add tests

Fixed issue where no nodes were intersecting

-created new issue where every node is now intersecting

Commit:
Basic working algorithm for detecting intersecting mesh edges

-Given a mesh it iterates through each element and compares it to every other element
-For each pair of elements it compares their edges to see if any intersect
-Checks that edges don't share any nodes before checking for overlap
-Prints out where the intersections are and prints out how many intersecting edges were found
-Can be improved in future versions so that it doesn't have to loop through every element twice
-Tests still need to be added

refs #26596

Commit:
Added more checks to intersecing edges algorithm

-Now it checks if the bounding boxes of the two elements being checked intersect
-If they do not intersect then they can't have edges that intersect so they are skipped
-Moved up edge check so now before checking for overlap, it checks to see if the two edges have already been checked
-Removed commented out code, and changed certain variable types from double to auto
-console output is now done in MehBaseDiagnosticsUtils.C and prints the coordinates where the edges intersect

refs #26596

Commit:
Added intersecting edge test and added tolerance parameter

-New test that generates a mesh with intersecting edges and tests it
-Made tolerance parameter adjustable. User can change it by adding intersection_tol= to diagnostic portion of input file
-Changed some varible type to auto and made comments more specific

closes #26596

GiudGiud commented:
The conversion from hex to tet in moose should be safe. There's a heuristic based on node ids to prevent that iirc. 
However external meshes could have that problem

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2462875/) on 56a74f3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f5681a9d7aa9ca34b44531bf6331eb7ae4d24886`

DanielYankura commented:
> The conversion from hex to tet in moose should be safe. There's a heuristic based on node ids to prevent that iirc. However external meshes could have that problem

That's a fair point. For me to make a test mesh in MOOSE that had this issue I had to mesh two blocks separately, convert to tets, and then rotate the one block 180 degrees so that the edges no longer matched.

Commit:
Implemented style changes reccomended by Civet

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2463014/) on 449303b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28701/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2463029/) on 449303b wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>f5681a</th>
      <th colspan=4>#28701 449303</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.05%</td>
      <td>85.06%</td>
      <td>+0.01%</td>
      <td>98.31%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>105933</td>
      <td>106047</td>
      <td>+114</td>
      <td>116</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18625</td>
      <td>18628</td>
      <td>+3</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28701/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28701/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,107 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+      ((std::fabs(n1x - n4x) < intersection_tol) && (std::fabs(n1y - n4y) < intersection_tol) &&
+       (std::fabs(n1z - n4z) < intersection_tol)) ||
+      ((std::fabs(n2x - n3x) < intersection_tol) && (std::fabs(n2y - n3y) < intersection_tol) &&
+       (std::fabs(n2z - n3z) < intersection_tol)))
```
if the nodes overlap and are not the same

DanielYankura replied:
So the reason I added that check was because I figured that if two nodes had the same x,y,z coordinates then they were the same node. I did try just checking if n1==n3 or n1==n4, ect. But there were a couple of cases where it didn't recognize the two nodes as equal, but comparing their coordinates did work.

GiudGiud replied:
not necessarily in a broken mesh. we could have nodes that overlap but do not have the same id
checking the id is enough to check equality

DanielYankura replied:
So I think I figured out why using n1==n3 wasn't working before. I had a bug where it was assigning the same id to every node. I did notice though that if it's implemented this way that with unmerged meshes with multiple volumes it will count all touching edges as intersecting since their nodes have different ids. I don't think that this is necessarily a bad thing but I wanted to double check that that's how we want it to behave?

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,107 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+
+  if (std::fabs(denominator) < intersection_tol)
+  {
+    // This indicates that the intersecting line is vertical so they don't intersect
```
both lines are parallel ?

DanielYankura replied:
You're right, it should say that both lines are parallel.

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,107 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+  auto mua = numerator / denominator;
+  auto mub = (d1343 + (mua * d4321)) / d4343;
+
+  // Use these values to solve for the two poits that define the shortest line segment
```
```suggestion
  // Use these values to solve for the two points that define the shortest line segment
```

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,107 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+  auto nby = n3y + mub * (n4y - n3y);
+  auto nbz = n3z + mub * (n4z - n3z);
+
+  // This method assume the two lines are infinite. This check to make sure na and nb are part of
+  // their respective line segments
+  if ((nax < std::min(n1x, n2x)) || (nax > std::max(n1x, n2x)) || (nay < std::min(n1y, n2y)) ||
+      (nay > std::max(n1y, n2y)) || (naz < std::min(n1z, n2z)) || (naz > std::max(n1z, n2z)))
+  {
+    return false;
+  }
+
+  if ((nbx < std::min(n3x, n4x)) || (nax > std::max(n3x, n4x)) || (nby < std::min(n3y, n4y)) ||
+      (nay > std::max(n3y, n4y)) || (nbz < std::min(n3z, n4z)) || (naz > std::max(n3z, n4z)))
+  {
+    return false;
+  }
```
```suggestion
  // This method assume the two lines are infinite. This check to make sure na and nb are part of
  // their respective line segments
  if (mua < 0 || mua > 1)
    return false;
  if (mub < 0 || mub > 1)
    return false;
```

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,107 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+bool
+checkEdgeOverlap(const std::unique_ptr<Elem> & edge1,
+                 const std::unique_ptr<Elem> & edge2,
+                 const ConsoleStream & console,
```
I would remove the console here and instead return a reference to the intersection point

then you can have a good message in the MG with all the info you could want (element concerned, etc)

GiudGiud reviewed (changes requested):
looks good

DanielYankura commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1252,6 +1262,90 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+      std::vector<std::unique_ptr<Elem>> other_edges(other_elem->n_edges());
+      for (auto j : other_elem->edge_index_range())
+        other_edges[j] = other_elem->build_edge_ptr(j);
+      for (auto & edge : elem_edges)
+      {
+        for (auto & other_edge : other_edges)
```
So maybe there is a better way, but I used the reference because elem->build_edge_ptr() creates a unique pointer to the edge, so when I don't take the reference I get an error because it's trying to create a copy of a unique_ptr. If you know of a better way around this let me know!

GiudGiud replied:
ah no this is fine. I didnt see they were unique_ptrs

DanielYankura commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1252,6 +1262,90 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+          auto n4 = *node_list2[1];
+          // Check if the edges have already been added to our check_edges list
+          if (std::find(checked_edges.begin(), checked_edges.end(), n1) != checked_edges.end() &&
+              std::find(checked_edges.begin(), checked_edges.end(), n2) != checked_edges.end() &&
```
So dividing by 2 at the end would work in some cases but not all. If I don't have that check there then it will count the same edge for every cell that it's a part of so if could be counted twice or possibly more than that. If you want I could just take out the check and have it print out all that info. Or I could also turn checked_edges into a set instead. If I recall correctly finding elements in a std::set is O(log n) so it would at least scale better than searching through a vector of nodes?

GiudGiud replied:
checked_edges into a set is good

but actually it s not checked-edges ? it's more like overlapping_edges_nodes?
in which case complexity is not quadratic (unless the whole mesh is filled with overlaps, which we dont expect)

so the check can stay

Commit:
Made changes reccomended in PR

-To check if two edges should be checked for overlap, node ids are compared to see if they share any nodes
-checked_edges_nodes is now in a set
-_console error message is now in MeshDiagnosticsGenerator.C and includes element ids
-In test the two blocks are stitched so any touching nodes should be merged
-Other minor improvements to make code more readable

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2467343/) on 7f7aaa4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6ffbd76688c2c09567f930c89298b6e5a6a403b8`

Commit:
Style changes to pass Civet tests

Commit:
Added detail tests file

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,85 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+                 const Real intersection_tol)
+{
+  // Get nodes from the two edges
+  const Node * n1 = edge1->get_nodes()[0];
```
```suggestion
  const Node * const n1 = edge1->get_nodes()[0];
```

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,85 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+  const Node * n4 = edge2->get_nodes()[1];
+
+  // get x,y,z coordinates for each node
+  auto n1x = n1->operator()(0);
```
```suggestion
  const auto n1x = n1->operator()(0);
```

GiudGiud replied:
same everywhere in this routine

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,85 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+
+bool
+checkEdgeOverlap(const std::unique_ptr<Elem> & edge1,
+                 const std::unique_ptr<Elem> & edge2,
```
you need to check that these elements are actually edges. use a mooseAssert

DanielYankura replied:
I included a check that should return the element type and I have it fail if it's anything but 0. If we wanted something a bit more straightforward I can just have it check that get_nodes() returns 2 nodes? Let me know what you think.

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,85 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+  auto n4y = n4->operator()(1);
+  auto n4z = n4->operator()(2);
+
+  // Check that none of these points are the same
```
```suggestion
  // Check that the two edges are not sharing a node
```

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,85 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+
+  // Calculate distance between these two nodes
+  double distance =
+      std::sqrt(std::pow(nax - nbx, 2) + std::pow(nay - nby, 2) + std::pow(naz - nbz, 2));
```
use `Utility::pow<2>(` instead of std::pow, it s faster

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1252,6 +1262,99 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+    std::vector<std::unique_ptr<Elem>> elem_edges(elem->n_edges());
+    for (auto i : elem->edge_index_range())
+      elem_edges[i] = elem->build_edge_ptr(i);
+    for (auto other_elem : mesh->active_element_ptr_range())
```
```suggestion
    for (const auto other_elem : mesh->active_element_ptr_range())
```

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1252,6 +1262,99 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+
+      // Get bounding boxes for both elements. If the bounding boxes don't intersect then no edges
+      // will either
+      auto boundingBox1 = elem->loose_bounding_box();
+      auto boundingBox2 = other_elem->loose_bounding_box();
```
```suggestion
      const auto boundingBox1 = elem->loose_bounding_box();
      const auto boundingBox2 = other_elem->loose_bounding_box();
```

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1252,6 +1262,99 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+      }
+    }
+  }
+  diagnosticsLog("Number of intersecting edges: " + Moose::stringify(num_intersecting_edges),
```
```suggestion
  diagnosticsLog("Number of intersecting element edges: " + Moose::stringify(num_intersecting_edges),
```

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1252,6 +1262,99 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+          // Now compare edge with other_edge
+          std::vector<double> intersection_coords(3);
+          bool overlap = MeshBaseDiagnosticsUtils::checkEdgeOverlap(
+              edge, other_edge, intersection_coords, _non_matching_edge_tol);
```
this check only works for straight edges btw. EDGE3 will be an issue. add a check on the element type earlier

DanielYankura replied:
In MeshBaseDiagnosticsUtils I included a mooseAssert at the beginning that checks to see if they are edges. I had it check to make sure they are EDGE2 as well. If you think I should add a check earlier than this I can, but I don't think having the wrong edge type will cause any issues before that check?

GiudGiud replied:
the assert will not catch user inputs

the assert is great inside the utils, in this generator we ll want a check

GiudGiud replied:
in the generator, you can tell the user about the type of the parent element, which they ll understand better than edge

DanielYankura replied:
Sounds good, I'll add a check in the generator then.

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,85 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
 }
+
+bool
+checkEdgeOverlap(const std::unique_ptr<Elem> & edge1,
```
```suggestion
checkFirstOrderEdgeOverlap(const std::unique_ptr<Elem> & edge1,
```

Commit:
Made more changes for PR

-Changed most variable types to const auto
-Changed name of overlap check to checkFirstOrderEdgeOverlap
-Added mooseAssert to check that elements being checked are of type EDGE2
-Updated comments and console output message

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2470369/) on 84b561e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6ffbd76688c2c09567f930c89298b6e5a6a403b8`

Commit:
More style changes needed to pass Civet tests

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +66,95 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+                           const Real intersection_tol)
+{
+  // check that the two elements are of type EDGE2
+  mooseAssert(edge1->side_type(*(edge1->side_index_range()).end()) == 0,
```
why are you not using the element of edge1 and edge2?

DanielYankura replied:
So I might have totally missed something, but when I looked through the documentation for elements and edges I didn't see any member variables for what type of element it was?

GiudGiud replied:
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1Elem.html#ad06f8ac22978144723179eed42d10bae

DanielYankura replied:
That makes more sense. I think my mistake was that I saw it was pure virtual and didn't see where it was implemented. Anyways just fixed it!

Commit:
Fixed edge assertion to call edge->type()

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2470574/), step Clang format on e9f3fb6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6ffbd76688c2c09567f930c89298b6e5a6a403b8`

Commit:
Added edge check in MeshDiagnosticsGenerator

-If edges are not of type EDGE2 a warning is printed out with edge and cell type
-The edge intersection check is then skipped for that element

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2471881/), step Clang format on 59d9208 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a31ebecb4216bdedfbd1d6ad3ec6c94969c03c07`

Commit:
Changed edge check from type == 0 to type == EDGE2

DanielYankura commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1252,6 +1262,118 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+          }
+
+          //Check element/edge type 
+          if (edge->type() != 0)
```
I made the change in MeshDiagnosticsGenerator and MeshBaseDiagnosticsUtils!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2472013/), step Clang format on 14b1664 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a31ebecb4216bdedfbd1d6ad3ec6c94969c03c07`

Commit:
More style changes for Civet tests

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2472115/), step Docs: sync website on f40f6d2 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28701/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2472130/), step Generate coverage on f40f6d2 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>a31ebe</th>
      <th colspan=4>#28701 f40f6d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.05%</td>
      <td>85.05%</td>
      <td>+0.00%</td>
      <td>87.97%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>106006</td>
      <td>106123</td>
      <td>+117</td>
      <td>117</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18636</td>
      <td>18651</td>
      <td>+15</td>
      <td>16</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28701/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28701/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28701/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 87.97% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1252,6 +1262,122 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+                "Edge of type " + Utility::enum_to_string(edge->type()) + " was found in cell " +
+                std::to_string(elem->id()) + " which is of type " +
+                Utility::enum_to_string(elem->type()) + '\n' +
+                "The edge intersection check only works for EDGE2 elements";
+            _console << element_message << std::endl;
```
```suggestion
                "The edge intersection check only works for EDGE2 elements.\nThis message will not be output again";
            mooseDoOnce(_console << element_message << std::endl;)
```

GiudGiud replied:
otherwise it is going to be so so much output

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1252,6 +1262,122 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+          }
+          if (other_edge->type() != EDGE2)
+          {
+            std::string element_message =
+                "Edge of type " + Utility::enum_to_string(other_edge->type()) +
+                " was found in cell " + std::to_string(other_elem->id()) + " which is of type " +
+                Utility::enum_to_string(other_elem->type()) + '\n' +
+                "The edge intersection check only works for EDGE2 elements";
+            _console << element_message << std::endl;
```
```suggestion
```

Commit:
Modified wrong edge check so that it only prints message once

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1252,6 +1262,115 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+  if (!mesh->is_serial())
+    mooseError("Only serialized/replicated meshes are supported");
+  unsigned int num_intersecting_edges = 0;
+  std::set<Node> checked_edges_nodes;
```
```suggestion
  std::set<Node> overlapping_edges_nodes;
```
the "checked" part is what made me think we were storing every single node there

GiudGiud reviewed (commented):
looks good to me.
@roystgnr did you want to take a look ?

Commit:
changed checked_edges_nodes to overlapping_edges_nodes

roystgnr commented:
I'll take a quick pass now.

GiudGiud commented:
Let's avoid the merge commit and instead perform a rebase

DanielYankura commented:
I rebased and fixed the merge conflicts before pushing this time. That should fix the merge conflict errors this time.

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1321,6 +1330,115 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+          // Check element/edge type
+          if (edge->type() != EDGE2)
+          {
+            std::string element_message = "Edge of type " + Utility::enum_to_string(edge->type()) +
```
you're missing the include "enum_to_string.h" (or whatever it's really called) and that's why the non-unity build is failing

DanielYankura replied:
Thanks! I was trying to figure out why that was failing. I'll fix that right now.

DanielYankura committed:
added string_to_enum.h reference for Civet test

DanielYankura committed:
Included utility.h for civet test

DanielYankura committed:
changed overlapping nodes set to take an id instead of a node to avoid libmesh issues

roystgnr reviewed (changes requested):
I'm really sorry for letting this slip my mind for so long!

The one thing we've got to fix here is the false negatives on the count of intersecting edges, but we might as well make the easier changes to clean up the style while we're at it; no need to have six line chunks of code that can be replaced by one API call.

The one thing that's killing me here is that we've got an O(N^2) algorithm where O(N log N) should do, but the octree we have in libMesh isn't *quite* the right kind of octree for this purpose, and so our alternatives are either "make it the right kind of octree" (spending a ton of man-hours) or "use it anyway and accept a possibility of false negatives" (not just on intersection count but on *existence*, which is unacceptable if users manage to trigger it).  How fast does this currently run on a mesh with a couple million elements?  Maybe speeding it up by 2x is good enough for now and we can wait on asymptotically-faster until users with bigger meshes complain.

DanielYankura commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +67,93 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+      (n4x - n3x) * (n2x - n1x) + (n4y - n3y) * (n2y - n1y) + (n4z - n3z) * (n2z - n1z);
+  const auto d1321 =
+      (n1x - n3x) * (n2x - n1x) + (n1y - n3y) * (n2y - n1y) + (n1z - n3z) * (n2z - n1z);
+  const auto d4343 =
```
For that one it works out that way. I was thinking it may be better to keep it as is so that it's more clear that it's doing the same thing as the other 3 calculations?

DanielYankura replied:
Or I could rewrite it as a dot product like the rest?

roystgnr replied:
I'd still prefer the `norm_sq()` but a dot product would be fine for the symmetry.

DanielYankura committed:
Made changes to improve efficiency and eliminate risk of false positives/negatives

-Bounding boxes are now calculated in their own loop and added to a map
-This map gets called later when they are needed
-Intersecting edges are stored as a set of arrays with 4 nodes each
-This prevents false positives/negatives when nodes may be part of multiple intersecting edges
-Made changes to loop so it doesn't calculate distance between the same edges twice
-Replaced Node calculations with Point instead to improve readability

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2556369/), step Clang format on 091e0ae wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 37b999b88168e63b60d24edc34afeda04ee57b8a`

DanielYankura committed:
style changes for Civet

DanielYankura commented:
So I did some experimenting to see how long this would take to run on larger meshes. I tested on one mesh that had 2 million cells and one with 1 million. Neither has finished so I know the answer is more than 2-3 hours. I'll try again with some slightly smaller meshes and see if I can get more useful results.

DanielYankura committed:
Removed utility.h because it is no longer needed for calculating distance

DanielYankura commented:
So this morning I started a test with 250k cells. It's still going so it's not looking great. Though I did get a chance to test out ones in the 10k range yesterday and those only took a couple of minutes. I'll get a test going with 100k cells and see if it takes unreasonably long.

roystgnr commented:
A couple minutes times 25² (10k→250k) would get us into the couple-dozen-hour range, yeah.  And then 1M elements would be half a week., and 100k would still be a few hours.  Maybe it's time to look into using the PointLocator here; a test that could theoretically miss failures in some uncommon cases is still better than a test nobody ever runs.

DanielYankura commented:
You mentioned earlier that to get it to O(N Log N) the octree for libmesh would have to be modified. Would we need to implement those changes to get PointLocator to work?

roystgnr commented:
To get it to work *reliably*, I think so.  Right now you can use a PointLocator to find all elements that intersect (contain, in this case) a single point (including all elements that almost contain it, to some relatively small tolerance), but you can't use it to find all elements that intersect a *line segment*, and we'd need that capability (or more realistically, a catch-all supercapability like "elements whose bounding boxes intersect an arbitrary bounding box") to make sure we caught every edge intersection for certain.  Using the PointLocator on both endpoints of an edge would probably catch 99.99% of edge intersection cases, but it would still be possible to have two edges intersect in such a way that neither edge is on any element that contains either of the other edge's vertices.  Imagine two very-skinny triangles overlapping at right angles to each other, for example.

But it might be worth just implementing the 99.99% case, if we don't have time to write a whole new set of PointLocator search algorithms and if the O(N²) algorithm is too slow for many users.

DanielYankura commented:
So I might be misunderstanding what you're saying but I'm not sure that the point locator would work for the test cases I have set up. I took two hex meshes and converted them to tets and then rotated one of them 180 and merged the two. So all nodes match up with another node on the opposite volume, but it still has edges that are intersecting. The issue you described sounds like something the nonconformal check would catch?

roystgnr commented:
Oh, I'm not talking about using the PointLocator to check whether nodes match up, I'm talking about using it to determine the set of elements to check.

Instead of
```
for (elem : every_element) // 1M iterations
  for (other : every_other_element) // 1M iterations
    check_edge_intersections(elem, other);
```

We'd have
```
for (elem : every_element) // 1M iterations
  for (other: every_other_element_nearby) // Maybe 150 iterations tops?
    check_edge_intersections(elem, other);
```

After using `PointLocator` queries with `elem` vertices (or elem edge vertices, if we reorder the loops, to get that 150 down a bit further) to determine the `every_other_element_nearby` set.  The edge intersection tests wouldn't change, we'd just run them a few thousand times less often on a big mesh.

DanielYankura commented:
That makes a lot more sense. I'll take a look at implementing that.

roystgnr commented:
Well, I hope it still makes more sense after we see the results.  PointLocator queries are quite a bit more expensive than bounding box intersection tests.  I'm just hoping they're not *so* much more expensive that they obviate the asymptotic analysis.  Ten million PointLocator calls from a million elements surely has to be faster than literally-a-trillion bounding box intersection calls, but there must be some point (10k locators from 1k elements, vs 1M intersections?) where a mesh is small enough that the locator code will actually be slower, and I'm just hoping that that point is already so fast that we won't care.

DanielYankura commented:
Sorry I meant to start a 100k test yesterday but I completely forgot. I did have a 250k test running on sawtooth for 8 or 9 hours though and it did not finish in the time I had. I can do some smaller tests today and see what it can finish in a few hours.

DanielYankura commented:
By the way my 100k cell test is still running but I started a 50k mesh test case and it took just under 2 hours. So 100k will probably take quite a while to finish.

DanielYankura committed:
Odd issue with merge conflicts. This should fix it

DanielYankura committed:
Whitespace issue fixed for civet

DanielYankura committed:
Improved efficiency of nonmatching edge check

-Should run in O(n log n)
-Nested for loop looks for nearby elements and loops through only those
-No longer had to check each element against every other element

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2577057/), step Clang format on acb1e18 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28701/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cedfd055b01b8ccebc50e5d10a89847caa8e24ae`

DanielYankura committed:
Style changes for Civet

DanielYankura commented:
So it looks like the only test it's failing is building FENIX, but I don't believe that that's related to anything in this PR?

loganharbour commented:
> So it looks like the only test it's failing is building FENIX, but I don't believe that that's related to anything in this PR?

I just invalidated. It's unrelated.

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1321,6 +1331,115 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+
+          // Check if the edges have already been added to our check_edges list
+          if (overlapping_edges_nodes.count(n1->id()) && overlapping_edges_nodes.count(n2->id()) &&
+              overlapping_edges_nodes.count(n3->id()) && overlapping_edges_nodes.count(n4->id()))
```
@DanielYankura this looks like the only unresolved item here?

DanielYankura replied:
I believe I implemented this change already? I think it was is this commit [Made changes to improve efficiency and eliminate risk of false positi…](https://github.com/idaholab/moose/pull/28701/commits/d275546c8348332304f40a2da38531fcab803f82)

GiudGiud replied:
ok I see now

GiudGiud commented:
So performance issues are resolved?

DanielYankura commented:
Yes, my last commit implemented Roy's suggestions for improving performance.

GiudGiud commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -66,4 +67,93 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
+  mooseAssert(edge1->type() == EDGE2, "Elements must be of type EDGE2");
+  mooseAssert(edge2->type() == EDGE2, "Elements must be of type EDGE2");
+  // Get nodes from the two edges
+  const Node * const n1 = edge1->get_nodes()[0];
```
resolving (mostly addressed)

roystgnr commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1429,26 +1430,28 @@ MeshDiagnosticsGenerator::checkNonMatchingEdges(const std::unique_ptr<MeshBase>
     bounding_box_map.insert({elem, boundingBox});
   }
 
+  std::unique_ptr<PointLocatorBase> point_locator = PointLocatorBase::build(TREE_ELEMENTS, *mesh);
```
Would `point_locator = mesh->sub_point_locator()` not work here?  In this use case I admit it's unlikely either that someone else has already constructed a point locator we can use or that someone else will take the opportunity to reuse the one we construct, but in general it's better to leave the point locator in the hands of the `Mesh` unless we need some particular non-default setting.

roystgnr commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -69,7 +69,7 @@ checkNonConformalMesh(const std::unique_ptr<MeshBase> & mesh,
 bool
 checkFirstOrderEdgeOverlap(const std::unique_ptr<Elem> & edge1,
-                           const std::unique_ptr<Elem> & edge2,
+                           const std::unique_ptr<const Elem> & edge2,
```
As long as you've got me looking here - is there a reason we want to pass const references to smart-pointers to elements here?  I'd think `const Elem & edge1, const Elem & edge2` would be the right signature to indicate that we're neither changing either edge element nor doing anything tricky with their memory management.

DanielYankura replied:
No sure why I had edges 2 as const but edge1 wasn't? Just fixed that in my latest commit.

roystgnr reviewed (changes requested):
Really happy to see the point locator idea worked!  How much speedup did you see?

DanielYankura committed:
Minor changes for PR

- point_locator is now initialized with sub_point_locator()
- edge1 is now const to match edge2

DanielYankura commented:
I don't recall the exact difference but I know previously a 100k cell mesh took somewhere in the ballpark of a 1-3 hours and now it can do it in about a minute.

loganharbour commented on framework/include/utils/MeshBaseDiagnosticsUtils.h:
```diff
@@ -25,7 +25,7 @@ void checkNonConformalMesh(const std::unique_ptr<libMesh::MeshBase> & mesh,
                            unsigned int & num_nonconformal_nodes);
 
-bool checkFirstOrderEdgeOverlap(const std::unique_ptr<Elem> & edge1,
+bool checkFirstOrderEdgeOverlap(const std::unique_ptr<const Elem> & edge1,
```
What Roy is alluding to is doing this instead

```
checkFirstOrderEdgeOverlap(const Elem & edge1,
                           const Elem & edge2,
```

DanielYankura replied:
That makes more sense. So the reason I had it set up that way is because `build_edge_ptr()` returns a `unique_ptr<Elem>` and the easiest method was to just pass that in. If you want I can probably make a copy that's just a const Elem or did you have something else in mind?

GiudGiud replied:
dereference the pointer to get a reference

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -1388,6 +1400,130 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
+  }
+
+  std::unique_ptr<PointLocatorBase> point_locator =
+      mesh->sub_point_locator(); // PointLocatorBase::build(TREE_ELEMENTS, *mesh);
```
```suggestion
      mesh->sub_point_locator();
```

DanielYankura committed:
Change for PR

-Removed commented out code
-checkFirstOrderEdgeOverlap now takes a const Elem instead of a uniqu_ptr
